apps/shipments/models.py

```diff
@@ -238,22 +271,61 @@ def get_device_request_url(self):
         LOG.debug(f'Getting device request url for device with vault_id {self.vault_id}')
         return f"{settings.PROFILES_URL}/api/v1/device/?on_shipment={self.vault_id}"
 
-    def update_vault_hash(self, vault_hash):
+    def update_vault_hash(self, vault_hash, rate_limit=False):
         LOG.debug(f'Updating vault hash {vault_hash}')
         async_job = None
         if self.load_data and self.load_data.shipment_id:
-            LOG.debug(f'Updating for shipment_id {self.load_data.shipment_id}')
-            async_job = AsyncJob.rpc_job_for_listener(
-                rpc_class=ShipmentRPCClient,
-                rpc_method=ShipmentRPCClient.update_vault_hash_transaction,
-                rpc_parameters=[self.shipper_wallet_id,
-                                self.load_data.shipment_id,
-                                '',
-                                vault_hash],
-                signing_wallet_id=self.shipper_wallet_id,
-                listener=self)
+            if rate_limit:
+                LOG.debug(f'Shipment {self.id} requested a rate-limited vault hash update')
+                job_queryset = AsyncJob.objects.filter(
+                    joblistener__shipments__id=self.id,
+                    state=JobState.PENDING,
+                    parameters__rpc_method=ShipmentRPCClient.update_vault_hash_transaction.__name__,
+                    parameters__signing_wallet_id=self.shipper_wallet_id,
+                )
+                if not job_queryset.count():
+                    LOG.debug(f'No pending vault hash updates for {self.id}, '
+                              f'sending one in {settings.VAULT_HASH_RATE_LIMIT} minutes')
+                    async_job = AsyncJob.rpc_job_for_listener(
+                        rpc_method=ShipmentRPCClient.update_vault_hash_transaction,
+                        rpc_parameters=[self.shipper_wallet_id,
+                                        self.load_data.shipment_id,
+                                        '',
+                                        vault_hash],
+                        signing_wallet_id=self.shipper_wallet_id,
+                        listener=self,
+                        delay=settings.VAULT_HASH_RATE_LIMIT)
+                else:
+                    # Update vault hash for current queued job
+                    async_job = job_queryset.first()
```

Comment:
It could be possible that there are multiple pending update_vault_hash jobs for a single shipment. (ex: one for add_tracking_data and one for add_shipment_data). This is fairly unlikely as add_shipment_data isn't a delayed task... but this case should be handled :+1: 

I am going to be updating all of the jobs - but this made me think more about related conditions and add a lock to the asyncjob task.